src/processpriv.h

```diff
@@ -56,22 +77,246 @@ class ProcessPriv : public AbstractPrivateData {
     DLLLOCAL static void waitForTermination(int pid, ExceptionSink* xsink);
 
 private:
-    bp::child *m_process;
-    bp::opstream m_in;
-    bp::ipstream m_out;
-    bp::ipstream m_err;
+    ResolvedCallReferenceNode* optsExecutor(const char* name, const QoreHashNode* opts, ExceptionSink* xsink);
 
-    const ResolvedCallReferenceNode* optsExecutor(const char * name, const QoreHashNode *opts, ExceptionSink *xsink);
+    bp::environment optsEnv(const QoreHashNode* opts, ExceptionSink* xsink);
+    const char* optsCwd(const QoreHashNode* opts, ExceptionSink* xsink);
 
-    bp::environment optsEnv(const QoreHashNode *opts, ExceptionSink *xsink);
-    const char* optsCwd(const QoreHashNode *opts, ExceptionSink *xsink);
+    //! Process exe arguments passed through constructor.
+    void processArgs(const QoreListNode* arguments, std::vector<std::string>& out);
+    
+    //! Prepare stdin ASIO buffer for use in async_write operation.
+    void prepareStdinBuffer();
 
 #ifdef __linux__
     DLLLOCAL static QoreHashNode* getMemorySummaryInfoLinux(int pid, ExceptionSink* xsink);
 #endif
 #if defined(__APPLE__) && defined(__MACH__)
     DLLLOCAL static QoreHashNode* getMemorySummaryInfoDarwin(int pid, ExceptionSink* xsink);
 #endif
+
+    //! Thread-safe input buffer. Used for writing to child process's stdin.
+    class InputBuffer {
+    public:
+        InputBuffer() = default;
+        ~InputBuffer() {
+            m_cv.notify_all();
+        }
+
+        //! Get size of data currently in buffer.
+        size_t size() {
+            return m_buf.size();
+        }
+
+        //! Write data into the input buffer.
+        void write(const char* src, size_t n) {
+            if (!src || !n)
+                return;
+
+            std::lock_guard<std::mutex> lock(m_mtx);
+            m_buf.append(src, n);
+        }
+
+        //! Extract data from the input buffer.
+        size_t extract(std::vector<char>& dest, size_t n) {
+            if (!n)
+                return 0;
+
+            std::lock_guard<std::mutex> lock(m_mtx);
+
+            // fix the read size
+            if (m_buf.size() < n)
+                n = m_buf.size();
+
+            dest.resize(n);
+            m_buf.copy(dest.data(), n, 0);
+            m_buf.erase(0, n);
+            return n;
+        }
+
+    private:
+        std::mutex m_mtx;
+        std::condition_variable m_cv;
```

Comment:
can u name it better then just `cv`? the type already says it's a `confition_variable` but for what it is used..